torch/_export/serde/serialize.py

```diff
@@ -636,45 +685,82 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
 class GraphModuleDeserializer:
     def __init__(self):
         self.serialized_name_to_node: Dict[str, torch.fx.Node] = {}
-        self.serialized_name_to_meta: Dict[str, Union[FakeTensor, int, torch.SymInt, torch.SymBool]] = {}
+        self.serialized_name_to_meta: Dict[str, Union[FakeTensor, int, torch.SymInt, bool, torch.SymBool]] = {}
         self.graph = torch.fx.Graph()
-        self.fake_tensor_mode = FakeTensorMode()
 
-    def deserialize_sym(
-        self, s: Union[SymInt, SymBool]
-    ) -> Union[int, torch.SymInt, bool, torch.SymBool]:
+    def deserialize_sym_int(self, s: SymInt) -> Union[int, torch.SymInt]:
         val = s.value
         if s.type == "as_symbol":
-            raise NotImplementedError("TODO(angelayi)")
+            expr, hint = val
+            sym = sympy.sympify(expr, locals=self.symbol_name_to_symbol)
+            return self.shape_env.create_symintnode(sym, hint=hint)
         elif s.type == "as_int":
             assert isinstance(val, int)
             return val
+        else:
+            raise SerializeError(
+                f"SymInt has invalid field type {s.type} with value {s.value}"
+            )
+
+    def deserialize_sym_bool(self, s: SymBool) -> Union[bool, torch.SymBool]:
+        val = s.value
+        if s.type == "as_symbol":
+            expr = sympy.sympify(val, locals=self.symbol_name_to_symbol)
+            return symbolic_shapes.SymBool(symbolic_shapes.SymNode(expr, self.shape_env, bool, None))
         elif s.type == "as_bool":
             assert isinstance(val, bool)
             return val
         else:
             raise SerializeError(
-                f"SymInt has invalid field type {s.type} with value {s.value}"
+                f"SymBool has invalid field type {s.type} with value {s.value}"
+            )
+
+    def deserialize_tensor_meta(
+        self,
+        tensor_meta: TensorMeta,
+        fake_tensor_mode: FakeTensorMode,
+    ) -> FakeTensor:
+        with fake_tensor_mode:
+            return cast(
+                FakeTensor,
+                torch.empty_strided(
+                    tuple(self.deserialize_sym_int(val) for val in tensor_meta.sizes),  # type: ignore[misc]
+                    tuple(self.deserialize_sym_int(val) for val in tensor_meta.strides),  # type: ignore[misc]
+                    device=deserialize_device(tensor_meta.device),
+                    dtype=_SERIALIZE_TO_TORCH_DTYPE[tensor_meta.dtype],
+                ),
             )
 
     def deserialize(
-        self, serialized_graph_module: GraphModule,
+        self,
+        serialized_graph_module: GraphModule,
+        shape_env: Optional[symbolic_shapes.ShapeEnv] = None,
```

Comment:
The `deserialize_range_constraints` function is called before deserializing the graph module, and I had it populate the shape_env with the symbols that appeared in those constraints first. 

This is under the assumption that most symbols that appear in the graph will also appear in the constraints (I want to say all symbols? but I'm not 100% sure on that although I don't have a counterexample). So I can first create all of the symbols in the shape_env, and then pass the `shape_env` and `symbol_name_to_symbol` list to the `GraphModuleDeserializer`. 

We need to populate `shape_env` with the ranges before tracing because if there's some data-dependent operation going on, like using a symint as a shape, deserializing the tensor metadata through the `torch.empty_strided` call will check on the ranges of the symint.

The `symbol_name_to_symbol` is also handy because when we're deserializing symint expressions, the `sympy.sympify` function uses it to map the same expressions to the same symbol (ex. if we have `symbol_name_to_symbol = {"s0", s0}`, whenever we see `"s0"` this will point to the same `s0`, and if we see something like `"7*s0"` this will also point to the same symbol `s0`.

Reply:
Ah, populated by side effect: missed that.

I think `constraints` only contain input dimension symbols, right? Whereas `shape_env` can contain data-dependent ones, I suppose.

I suppose `shape_env` better contain all symbols appearing in the graph. These should be a superset of things in `constraints`, I think. And `symbol_name_to_symbol` should have all symbols in the graph as well...?

Is there a way to assert on this?

Also, I still don't fully understand whether you actually expect external users to call graph module deserializer directly. Why not make it private, remove the optional / default part, etc.?